### PR TITLE
Move from 2.1 to 3.1

### PR DESCRIPTION
**Pull request by ChadNedzlek**
This is 100% going to break master, because many of the breaking changes from 2.1 to 3.1 only appear at runtime, and it's not possible to hit all of the code locally, especially when it comes to database interactions.

Unfortunately, because of the scope and importance of these changes (other people are blocked waiting for these changes), I believe this change should be exempt from the default build fix of "back out the changes", and we should do the work to forward fix it, all hands on deck.  I will likely be reaching out to people to help me diagnose strange problems in strange pieces of code I don't know a lot about.

### CODE REVIEW

File: src/DotNet.Status.Web/Startup.cs [hunk 1]
```diff
@@ -188,10 +200,10 @@ private void AddServices(IServiceCollection services)
             services.AddScoped<SimpleSigninMiddleware>();
             services.AddGitHubTokenProvider();
             services.AddSingleton<IInstallationLookup, InMemoryCacheInstallationLookup>();
-            services.AddContextAwareAuthenticationScheme(o =>
-            {
-                o.SelectScheme = p => p.StartsWithSegments("/api") ? "github-token" : IdentityConstants.ApplicationScheme;
-            });
+            //services.AddContextAwareAuthenticationScheme(o =>
```
Comment (missymessa) [inline src/DotNet.Status.Web/Startup.cs:203]: Keeping commented out code?

Comment (ChadNedzlek) [reply src/DotNet.Status.Web/Startup.cs:203]: :hang_head_low_emoji: No... I saw that several times and thought "I should deleted that"...

File: eng/Packages.props [hunk 2]
```diff
@@ -0,0 +1,110 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project>
+  <PropertyGroup>
+    <DotNetAssemblyVersion>3.1.3</DotNetAssemblyVersion>
+  </PropertyGroup>
+  <ItemGroup> 
+    <PackageReference Update="Azure.Core" Version="1.0.1" />
```
Comment (JohnTortugo) [inline eng/Packages.props:7]: Can we use version 1.0.2 here? https://github.com/dotnet/arcade-services/pull/1163

File: src/DotNet.Status.Web/Startup.cs (hunk 1 above)
Comment (ChadNedzlek) [reply src/DotNet.Status.Web/Startup.cs]: Done!

File: eng/Packages.props (hunk 2 above)
Comment (ChadNedzlek) [reply eng/Packages.props:7]: Sure. :-)

Comment (ChadNedzlek) [reply eng/Packages.props]: Done.

Comment (ChadNedzlek) [reply eng/Packages.props]: on the plus side, the usage of Packages.props should aleviate a ton of the package versin conflict. Since it's now not possible for any package in either arcade-services or Helix to get out sync.  (There is a tool in this PR to sync Package.props files between repos, I don't mess around when I go for consistenct)

File: src/DotNet.Status.Web/Program.cs [hunk 3]
```diff
@@ -7,35 +7,37 @@
...
     {
         public static void Main(string[] args)
         {
-            new WebHostBuilder()
-                .UseKestrel()
+            Host.CreateDefaultBuilder()
+                .ConfigureWebHostDefaults(host =>
+                {
+                    host.UseStartup<Startup>()
+                        .UseUrls("http://localhost:5000/")
```
Comment (sunandabalu) [inline src/DotNet.Status.Web/Program.cs:23]: Is this intended to be localhost url?

File: src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs [hunk 4]
```diff
@@ -72,12 +73,43 @@ public class BuildAssetRegistryContext : IdentityDbContext<ApplicationUser, Iden
...
+                {
+                    SubscriptionId = u.SubscriptionId,
+                    Action = u.Action,
+                    Success = u.Success,
+                    ErrorMessage = u.ErrorMessage,
+                    Method = u.Method,
+                    Arguments = u.Arguments,
+                    Timestamp = EF.Property<DateTime>(u, "SysStartTime")
+                });
+
```
Comment (sunandabalu) [inline src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:112]: How are these changes related to the version upgrade to 3.1?

File: src/Maestro/Microsoft.AspNetCore.ApiVersioning/ApiVersioningOptions.cs [hunk 5]
```diff
@@ -16,12 +16,14 @@ public class ApiVersioningOptions
         public Func<TypeInfo, string> GetName { get; set; } = DefaultGetName;
         public IVersioningScheme VersioningScheme { get; set; } = new PathVersioningScheme();
 
+        [CanBeNull]
         public static string DefaultGetVersion(TypeInfo controllerType)
```
Comment (sunandabalu) [inline src/Maestro/Microsoft.AspNetCore.ApiVersioning/ApiVersioningOptions.cs:20]: DefaultVersion can be null?

File: eng/common/internal/Tools.csproj [hunk 6]
```diff
@@ -2,14 +2,13 @@
...
-    <TargetFramework>net472</TargetFramework>
     <ImportDirectoryBuildTargets>false</ImportDirectoryBuildTargets>
   </PropertyGroup>
   <ItemGroup>
     <!-- Clear references, the SDK may add some depending on UsuingToolXxx settings, but we only want to restore the following -->
     <PackageReference Remove="@(PackageReference)"/>
-    <PackageReference Include="Microsoft.DotNet.IBCMerge" Version="$(MicrosoftDotNetIBCMergeVersion)" Condition="'$(UsingToolIbcOptimization)' == 'true'" />
-    <PackageReference Include="Drop.App" Version="$(DropAppVersion)" ExcludeAssets="all" Condition="'$(UsingToolVisualStudioIbcTraining)' == 'true'"/>
+    <PackageReference Include="Microsoft.DotNet.IBCMerge" Condition="'$(UsingToolIbcOptimization)' == 'true'" />
+    <PackageReference Include="Drop.App" ExcludeAssets="all" Condition="'$(UsingToolVisualStudioIbcTraining)' == 'true'"/>
```
Comment (alexperovich) [inline eng/common/internal/Tools.csproj:11]: This change will just get overriden by the the next arcade update. These properties should stick around.

Comment (ChadNedzlek) [reply eng/common/internal/Tools.csproj:11]: I guess as long as nothing builds this project it will be fine... Why is there a csproj in the common???

Comment (alexperovich) [reply eng/common/internal/Tools.csproj:11]: I have no idea.

File: src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs (hunk 4 above)
Comment (ChadNedzlek) [reply src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs:112]: The previous thing, DbQuery, isn't around anymore.  All of these changes are because of breaking changes because of standardizing from 2.1 to 3.1.  There are no functional changes anywhere that weren't necessitated by the change.

File: src/DotNet.Status.Web/Program.cs (hunk 3 above)
Comment (ChadNedzlek) [reply src/DotNet.Status.Web/Program.cs:23]: This is just moving line 37 to here.  The three methods don't work on the "CreateDefaultBuilder", they have go inside the "ConfigureWebHostDefaults" now.

File: src/Maestro/Microsoft.AspNetCore.ApiVersioning/ApiVersioningOptions.cs (hunk 5 above)
Comment (ChadNedzlek) [reply src/Maestro/Microsoft.AspNetCore.ApiVersioning/ApiVersioningOptions.cs:20]: This is a change I pulled in from @alexperovich. Alex?

File: src/Maestro/Microsoft.AspNetCore.ApiVersioning.Swashbuckle/swagger.json [hunk 7]
```diff
@@ -0,0 +1,3649 @@
+{
```
Comment (alexperovich) [inline src/Maestro/Microsoft.AspNetCore.ApiVersioning.Swashbuckle/swagger.json:1]: Remove this file please. I added it to my change by accident.

File: src/Maestro/Microsoft.AspNetCore.ApiVersioning/ApiVersioningOptions.cs (hunk 5 above)
Comment (alexperovich) [reply src/Maestro/Microsoft.AspNetCore.ApiVersioning/ApiVersioningOptions.cs:20]: The "version" of a controller can be null. If it is an unversioned controller then it has no version.

File: eng/common/internal/Tools.csproj (hunk 6 above)
Comment (ChadNedzlek) [reply eng/common/internal/Tools.csproj]: Reverted

File: src/Maestro/Microsoft.AspNetCore.ApiVersioning.Swashbuckle/swagger.json (hunk 7 above)
Comment (ChadNedzlek) [reply src/Maestro/Microsoft.AspNetCore.ApiVersioning.Swashbuckle/swagger.json]: Done.

File: src/Maestro/Microsoft.AspNetCore.ApiVersioning/ApiVersioningOptions.cs (hunk 5 above)
Comment (ChadNedzlek) [reply src/Maestro/Microsoft.AspNetCore.ApiVersioning/ApiVersioningOptions.cs:20]: Answered! I guess, yeah, like the "login" controller and stuff aren't versioned.

File: src/Maestro/Client/src/Microsoft.DotNet.Maestro.Client.csproj [hunk 8]
```diff
@@ -12,12 +11,10 @@
...
   <ItemGroup>
-    <PackageReference Include="Azure.Core" Version="1.0.2" />
-    <PackageReference Include="Newtonsoft.Json" Version="$(NewtonsoftJsonVersion)" />
-    <PackageReference Include="System.Collections.Immutable" Version="$(SystemCollectionsImmutableVersion)" />
-    <PackageReference Include="System.Text.Encodings.Web" Version="4.5.0" />
+    <PackageReference Include="Azure.Core" />
+    <PackageReference Include="Newtonsoft.Json" />
 
-    <PackageReference Include="Microsoft.DotNet.SwaggerGenerator.MSBuild" Version="$(MicrosoftDotNetSwaggerGeneratorMSBuildVersion)" PrivateAssets="all" />
+    <PackageReference Include="Microsoft.DotNet.SwaggerGenerator.MSBuild" PrivateAssets="all" />
```
Comment (alexperovich) [inline src/Maestro/Client/src/Microsoft.DotNet.Maestro.Client.csproj:17]: ><PackageReference Include="Microsoft.DotNet.SwaggerGenerator.MSBuild" PrivateAssets="all" /> [](start = 4, length = 92)

Can we put this `PrivateAssets="all"` in the Packages.props for this. This package should never be used without this.

File: src/Maestro/Maestro.Web/Maestro.Web.csproj [hunk 9]
```diff
@@ -19,44 +15,23 @@
...
-  </ItemGroup>
-
-  <ItemGroup>
-    <DotNetCliToolReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Tools" Version="2.1.0-preview1-final" PrivateAssets="All" />
+    <PackageReference Include="Azure.Identity" />
+    <PackageReference Include="EntityFrameworkCore.Triggers" />
+    <PackageReference Include="FluentValidation.AspNetCore" />
+    <PackageReference Include="Humanizer.Core" /> 
+    <PackageReference Include="Microsoft.AspNetCore.DataProtection.AzureKeyVault" />
+    <PackageReference Include="Microsoft.AspNetCore.DataProtection.Azurestorage" />
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/Maestro.Web.csproj:23]: >Azurestorage [](start = 67, length = 12)

Does this package actually have a lowercase `s`? Does this work this way?

File: src/Maestro/Client/src/Microsoft.DotNet.Maestro.Client.csproj (hunk 8 above)
Comment (ChadNedzlek) [reply src/Maestro/Client/src/Microsoft.DotNet.Maestro.Client.csproj:17]: Sure. :-)

File: src/Maestro/Maestro.Web/Startup.cs [hunk 10]
```diff
@@ -277,21 +292,52 @@ private void ConfigureApi(IApplicationBuilder app)
         {
             app.UseExceptionHandler(ConfigureApiExceptions);
 
-            app.UseAuthentication();
+            app.UseSwagger(
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/Startup.cs:295]: >app.UseSwagger( [](start = 12, length = 15)

This `UseSwagger` is too high. If you want to move this here the handler directly above it that forwards `/api/swagger.json` to `/api/{latestVersion}/swagger.sjon` needs to come before this one.

File: src/Maestro/Maestro.Web/Maestro.Web.csproj (hunk 9 above)
Comment (ChadNedzlek) [reply src/Maestro/Maestro.Web/Maestro.Web.csproj:23]: It doesn't actually have it. It builds though, which means the version is propigating through. I'll fix it anyway though. ;-)

File: src/Maestro/Maestro.Web/Startup.cs (hunk 10 above)
Comment (ChadNedzlek) [reply src/Maestro/Maestro.Web/Startup.cs:295]: Moved the other thing up (it needs to come before the Routing/Endpoints stuff.

File: src/Maestro/Maestro.Web/.config/settings.Development.json [hunk 11]
```diff
@@ -3,5 +3,6 @@
   "AppConfigurationUri": "https://maestrolocal.azconfig.io/",
   "BuildAssetRegistry": {
     "ConnectionString": "Data Source=localhost\\SQLEXPRESS;Initial Catalog=BuildAssetRegistry;Integrated Security=true"
-  }
+  },
+  "ForceLocalApi": true
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/.config/settings.Development.json:7]: This will break the local dev scenario where `ng serve` + `dotnet run` are combined to debug the angular bits.

File: src/Maestro/Microsoft.AspNetCore.ApiVersioning/Microsoft.AspNetCore.ApiVersioning.csproj [hunk 12]
```diff
@@ -1,18 +1,13 @@
...
   <ItemGroup>
-    <PackageReference Include="Microsoft.AspNetCore.Mvc.Abstractions" Version="2.1.1" />
-    <PackageReference Include="Microsoft.AspNetCore.Mvc.ApiExplorer" Version="2.1.1" />
-    <PackageReference Include="Microsoft.AspNetCore.Mvc.Core" Version="2.1.1" />
-    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="2.1.1" />
-    <PackageReference Include="Microsoft.Extensions.Options" Version="2.1.1" />
-    <PackageReference Include="JetBrains.Annotations" Version="11.1.0" PrivateAssets="All" />
+    <FrameworkReference Include="Microsoft.AspNetCore.App"/>
+
+    <PackageReference Include="JetBrains.Annotations" PrivateAssets="All" />
```
Comment (alexperovich) [inline src/Maestro/Microsoft.AspNetCore.ApiVersioning/Microsoft.AspNetCore.ApiVersioning.csproj:10]: I think this package should also have the PrivateAssets bit in the Packages.props

File: src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+
+            return (collapsedRemotePackages, errors);
+        }
+
+        private static async Task<Dictionary<string, Dictionary<string, string>>> FetchRemotePackageVersions(
+            Dictionary<string, Dependency> remoteDependencies)
+        {
+            var assembly = Assembly.GetEntryAssembly();
+            var client = new GitHubClient(new ProductHeaderValue(
+                "Microsoft.DotNet.Internal.Tools.SynchronizePackageProps",
```
Comment (alexperovich) [inline src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs:232]: NIT: `assembly.FullName`

File: src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+                assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion ??
+                assembly.GetName().Version.ToString()));
+
+            var
```
Comment (alexperovich) [inline src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs:236]: NIT: just put this on the same line.

File: src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs [hunk 15]
```diff
@@ -0,0 +1,397 @@
...
+                        WriteWarning($"{repoName} Packages.prop has Conditional PropertyGroup, which is not supported");
+                        continue;
+                    }
+
+                    foreach (XElement prop in propGroup.Elements())
+                    {
+                        if (prop.Attribute("Condition") != null)
+                        {
+                            WriteWarning(
+                                $"{repoName} Packages.prop property {prop.Name.LocalName} has Conditional PropertyGroup, which is not supported");
```
Comment (alexperovich) [inline src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs:308]: >PropertyGroup [](start = 106, length = 13)

NIT: this is an error for a Conditional property, not a conditional property group.

File: src/Maestro/Maestro.Web/.config/settings.Development.json (hunk 11 above)
Comment (ChadNedzlek) [reply src/Maestro/Maestro.Web/.config/settings.Development.json:7]: Oh crap, meant to leave this false. :-)

File: src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs [hunk 16]
```diff
@@ -0,0 +1,397 @@
...
+        {
+            Console.ForegroundColor = ConsoleColor.Yellow;
+            Console.Error.WriteLine(message);
+            Console.ResetColor();
+        }
+
+        [Flags]
+        private enum Errors
+        {
+            BadArguments = 1,
```
Comment (alexperovich) [inline src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs:373]: Why is the Flags but not all the values are single bits? BadLocalFormat is BadArguments | MissingFile. Is that intentional?

Comment (ChadNedzlek) [reply src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs:373]: The first nibble is for "you called me wrong".  Once we've decided you called it right, we can detect a lot of errors at once, so figured it was helpful to just mush all those together. :-)

File: src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs (hunk 13 above)
Comment (ChadNedzlek) [reply src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs:232]: Well, "Name". Full name includes all the version crap, which isn't legal in a product header. :-)

File: src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs (hunk 14 above)
Comment (ChadNedzlek) [reply src/Microsoft.DotNet.Internal.Tools.SynchronizePackageProps/Program.cs:236]: Weird that ReSharper did that. Fixed!

**Comment by alexperovich** [comment]
>                 return $"{baseName}<{string.Join(",", type.GetGenericArguments().Select(GetDisplayName))}>";

NIT: ", "

---
Refers to: src/Shared/Microsoft.DotNet.Internal.DependencyInjection.Testing/DependencyInjectionValidation.cs:168 in 9442603. [](commit_id = 9442603765ef36cf3bbd6aa4e2fa0e13bdd0ded0, deletion_comment = False)

File: src/Shared/Microsoft.DotNet.Internal.DependencyInjection.Testing.Tests/Dependencies.cs [hunk 17]
```diff
@@ -0,0 +1,131 @@
+namespace Microsoft.DotNet.Internal.DependencyInjection.Testing.Tests
```
Comment (alexperovich) [inline src/Shared/Microsoft.DotNet.Internal.DependencyInjection.Testing.Tests/Dependencies.cs:1]: Glad the tests have tests so we can test that the tests test the tests correctly, :P

Comment (ChadNedzlek) [reply src/Shared/Microsoft.DotNet.Internal.DependencyInjection.Testing.Tests/Dependencies.cs:1]: Indeed.  The logic to check the scoping is weird and complicated, so I wanted a way to make sure we detected it without having to change bad code. :-) In general, testing tests is silly, but these tests are hundreds and hundreds of lines of code, so needed some simple validation. :-)

File: src/Shared/Microsoft.DotNet.Web.Authentication/DefaultAuthorizeActionModelConvention.cs [hunk 18]
```diff
@@ -22,6 +23,13 @@ public DefaultAuthorizeActionModelConvention(string policyName)
 
         public void Apply(ActionModel action)
         {
+            // ASP.NET 3.1 broke IAllowAnonymousFilter, just find the attribute ourselves
+            // Otherwise it takes like 500 lines of code
+            if (action.ActionMethod?.GetCustomAttributes(true).Any(a => a is IAllowAnonymous || a is IAuthorizeData) ?? false)
```
Comment (alexperovich) [inline src/Shared/Microsoft.DotNet.Web.Authentication/DefaultAuthorizeActionModelConvention.cs:28]: Unfortunately this won't work if we but `[AllowAnonymous]` on the controller level, but we just shouldn't do that.

Comment (ChadNedzlek) [reply src/Shared/Microsoft.DotNet.Web.Authentication/DefaultAuthorizeActionModelConvention.cs:28]: K.  There is some code later that reads the "Controller.Attributes" already, so I _think_ that should be fine already... it wasn't reading the filters from the controller, but the attributes themselves.

**Comment by ChadNedzlek** [comment]
It's going in... wish us all luck!
